src/coreclr/System.Private.CoreLib/src/System/Threading/SyncTable.CoreCLR.cs

```diff
@@ -0,0 +1,34 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    internal static partial class SyncTable
+    {
+        internal static unsafe int AssignEntry(object obj, int* pHeader)
+        {
+            int slot = AssignEntryInternal(ObjectHandleOnStack.Create(ref obj));
+#if DEBUG
+            bool hasSlot = ObjectHeader.GetSyncEntryIndex(*pHeader, out int slotInHeader);
+            Debug.Assert(hasSlot, "Expected the header to have a sync entry index.");
+            Debug.Assert(slotInHeader == slot, $"Expected the slot in the header ({slotInHeader}) to match the assigned slot ({slot}).");
+#endif
+            return slot;
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "SyncTable_AssignEntry")]
+        private static partial int AssignEntryInternal(ObjectHandleOnStack obj);
```

Comment:
This will likely need to have fast path FCall / slow path QCall split to avoid perf regressions.